File: backend/app/models/gdd.py
from datetime import datetime
from sqlalchemy import (
    String,
    Date,
    DateTime,
    Float,
    Enum,
    Boolean,
    ForeignKey,
    UniqueConstraint,
    Column,
    Integer,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.core.database import Base
import enum


class TempUnit(enum.Enum):
    C = "C"
    F = "F"


class ResetType(enum.Enum):
    manual = "manual"
    threshold = "threshold"
    initial = "initial"
    application = "application"


class GDDModel(Base):
    __tablename__ = "gdd_models"
    __table_args__ = (UniqueConstraint("lawn_id", "name", name="uix_lawn_name"),)

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    lawn_id: Mapped[int] = mapped_column(ForeignKey("lawns.id"), nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    base_temp: Mapped[float] = mapped_column(Float, nullable=False)
    unit: Mapped[TempUnit] = mapped_column(Enum(TempUnit), nullable=False)
    start_date: Mapped[Date] = mapped_column(Date, nullable=False)
    threshold: Mapped[float] = mapped_column(Float, nullable=False)
    reset_on_threshold: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow
    )

    lawn = relationship("Lawn", back_populates="gdd_models")
    gdd_values = relationship(
        "GDDValue", back_populates="gdd_model", cascade="all, delete-orphan"
    )
    resets = relationship(
        "GDDReset", back_populates="gdd_model", cascade="all, delete-orphan"
    )
    parameter_history = relationship(
        "GDDModelParameters", back_populates="gdd_model", cascade="all, delete-orphan"
    )


class GDDModelParameters(Base):
    __tablename__ = "gdd_model_parameters"
    __table_args__ = (
        UniqueConstraint(
            "gdd_model_id", "effective_from", name="uix_gdd_model_effective_from"
        ),
    )

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    gdd_model_id: Mapped[int] = mapped_column(
        ForeignKey("gdd_models.id"), nullable=False
    )
    base_temp: Mapped[float] = mapped_column(Float, nullable=False)
    threshold: Mapped[float] = mapped_column(Float, nullable=False)
    reset_on_threshold: Mapped[bool] = mapped_column(Boolean, nullable=False)
    effective_from: Mapped[Date] = mapped_column(Date, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow
    )

    gdd_model = relationship("GDDModel", back_populates="parameter_history")


class GDDValue(Base):
    __tablename__ = "gdd_values"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    gdd_model_id: Mapped[int] = mapped_column(
        ForeignKey("gdd_models.id"), nullable=False
    )
    date: Mapped[Date] = mapped_column(Date, nullable=False)
    daily_gdd: Mapped[float] = mapped_column(Float, nullable=False)
    cumulative_gdd: Mapped[float] = mapped_column(Float, nullable=False)
    is_forecast: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    run: Mapped[int] = mapped_column(default=1, nullable=False)

    gdd_model = relationship("GDDModel", back_populates="gdd_values")


class GDDReset(Base):
    __tablename__ = "gdd_resets"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    gdd_model_id: Mapped[int] = mapped_column(
        ForeignKey("gdd_models.id"), nullable=False
    )
    reset_date: Mapped[Date] = mapped_column(Date, nullable=False)
    run_number: Mapped[int] = mapped_column(Integer, nullable=False)
    reset_type: Mapped[ResetType] = mapped_column(Enum(ResetType), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow
    )

    gdd_model = relationship("GDDModel", back_populates="resets")

File: frontend/src/types/gdd.ts
export interface GDDModel {
  id: number;
  name: string;
  base_temp: number;
  unit: "C" | "F";
  start_date: string;
  threshold: number;
  reset_on_threshold: boolean;
  location_id: number;
  created_at: string;
  updated_at: string;
  location?: Location;
}

export interface GDDValue {
  id: number;
  gdd_model_id: number;
  date: string;
  daily_gdd: number;
  cumulative_gdd: number;
  run: number;
  created_at: string;
}

export interface GDDReset {
  id: number;
  gdd_model_id: number;
  reset_date: string;
  reset_type: "manual" | "threshold";
  run_number: number;
  created_at: string;
}

export interface GDDParameterHistory {
  id: number;
  gdd_model_id: number;
  base_temp: number;
  threshold: number;
  reset_on_threshold: boolean;
  effective_from: string;
  created_at: string;
}

export interface GDDFormValues {
  name: string;
  base_temp: string;
  unit: "C" | "F";
  start_date: string;
  threshold: string;
  reset_on_threshold: boolean;
}

export interface GDDParameterEditFormValues {
  base_temp: string;
  threshold: string;
  reset_on_threshold: boolean;
  recalculate_history: boolean;
  effective_from: string;
  replace_all_history: boolean;
}

export interface GDDModelDashboard {
  id: number;
  location_id: number;
  name: string;
  base_temp: number;
  unit: string;
  threshold: number;
  created_at: string;
  updated_at: string;
  current_gdd?: number;
  last_reset?: string;
  run_number?: number;
}

// Import Location type from location.ts
import { Location } from "./location";
